Topics/Tech_Stacks/Writing_Clean_React_Code.md

```diff
@@ -0,0 +1,153 @@
+
+- **State Variables and Setters**: State variables should be prefixed with _is_, _has_, or _should_ to denote boolean values and the setters should be prefixed with _set_ followed by the name of the variable. For example, a state variable that stores whether a user is logged in should be named `isLoggedIn` and its setter should be named `setIsLoggedIn`.
+
+- **Event Handlers**: Event handlers should be named using camelCase and prefixed with _handle_. For example, a function that handles a button click should be named `handleClick` instead of `onClick`.`
+
+- **Props**: Props should be named using camelCase, similar to variables. Moreover, it is good practice to list props out when defining a component. This helps developers quickly identify the props that a component accepts. For example, a component that renders a button should be defined as follows:
+
+  ```jsx
+  const Button = ({ text, onClick }) => {
+    return <button onClick={onClick}>{text}</button>;
+  };
+  ```
+
+Notice that the onClick handler isn't defined in the component, and prop name isn't prefixed by _handle_. This is because the component is only responsible for rendering the button and not handling the button click. The onClick handler should be defined in the parent component that uses the Button component, and _in the parent component_ the event handler that is passed down to the Button component should be prefixed by _handle_. Like in the below example:
+
+```jsx
+const ParentComponent = () => {
+  const handleClick = () => {
+    // handle button click
+  };
+  return <Button text="Click Me" onClick={handleClick} />;
+};
+```
+
+- **Constants**: Constants should be named using uppercase letters and underscores to separate words. For example, a constant that stores user emails should be named `USER_EMAILS` instead of `user-emails` or `userEmails`.
+
+- **File Names**: File names should be named using PascalCase. For example, a file that contains a component that renders a navbar element should be named `NavbarElement.jsx` instead of `navbarElement.jsx` or `navbar-element.jsx`.
+
+- **CSS Classes and Files**: CSS classes should be named using kebab-case or camelCase. For example, a class that styles a div that acts as a container for a cards should be named `cards-container` or `cardsContainer` instead of `cards_container`. CSS Files follows the same convention as CSS classes.
+
+As to what you should name your classes, a popular convention is to use the BEM (Block, Element, Modifier) naming convention. This convention is useful for naming classes that are used to style components. It is a pretty big topic in its own right, so to learn more you might want to check out the [getbem.com website](https://getbem.com/introduction/).
```

Comment:
It's good to link a more detailed article but maybe just including 1 example here would be helpful.

Reply:
I agree. I will add an example.